### PR TITLE
Refactor mam_amicphys_1subarea_cloudy/clear

### PR DESCRIPTION
**Pull request by huiwanpnnl**
The original MAM microphysics code had two bottom-level driver subroutines, `mam_amicphys_1subarea_cloudy` and `mam_amicphys_1subarea_clear`, in which the various microphysics processes (e.g., nucleation, renaming) were called. These two subroutines were very (>80%) similar, although some calculations were done only for the cloudy subarea and some others were done only for the clear subarea.

In this PR, I have
- merged the two subroutines into one, now called `mam_amicphys_1subarea`,
- simplified the logic (e.g., the use of logical variables, if-statements, and temporary variables),
- did the usual clean-up (e.g., removed unused variables, added comments, etc.)

Note that the new subroutine `mam_amicphys_1subarea` is temporarily placed in a separate file `1subarea.F90` and imported into `modal_aero_amicphys.F90` using
```
#include "1subarea.F90"
```
This was because `modal_aero_amicphys.F90` is a very large file and I wanted to avoid potential conflict between different refactoring efforts.

### CODE REVIEW

File: components/eam/src/chemistry/modal_aero/1subarea.F90 [hunk 1]
```diff
@@ -0,0 +1,419 @@
+
+!--------------------------------------------------------------------------------
+! Call aerosol microphysics processes for a single (cloudy or clear) subarea
+! (with indices = lchnk,i,k,jsubarea)
+!
+! qgas3, qaer3, qaercw3, qnum3, qnumcw3 are the current incoming TMRs
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/1subarea.F90:6]: Maybe add units of TMRs in this comment.

File: components/eam/src/chemistry/modal_aero/1subarea.F90 [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+!    growth from smaller to larger modes (renaming) following cloud chemistry
+!    so gas TMRs are not changed
+!  - when do_cond = true, this routine also calculates changes involving
+!    gas-aerosol exchange (condensation/evaporation)
+!  - transfer of particles from hydrophobic modes to hydrophilic modes (aging)
+!       due to condensation
+! Currently, in a cloudy subarea, this routine does not do
+!  - new particle nucleation - because h2so4 gas conc. should be very low in cloudy air
+!  - coagulation - because cloud-borne aerosol would need to be included
+!--------------------------------------------------------------------------------
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/1subarea.F90:28]: Very nice summary!

File: components/eam/src/chemistry/modal_aero/1subarea.F90 [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+subroutine mam_amicphys_1subarea(&
+                                 do_cond_sub,            do_rename_sub,      &
+                                 do_newnuc_sub,          do_coag_sub,        &
+                                 nstep,      lchnk,      i,        k,        &
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/1subarea.F90]: If it is not too much work, would you please change `i` to `icol` (or `ii`) and `k` to `kk`?

File: components/eam/src/chemistry/modal_aero/1subarea.F90 [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+   use modal_aero_amicphys_control, only: r8, ntot_poaspec=>npoa, ntot_soaspec=>nsoa &
+                                        , max_mode, max_gas, max_aer &
+                                        , misc_vars_aa_type &
+                                        , gaexch_h2so4_uptake_optaa, ngas, igas_h2so4, igas_nh3 &
+                                        , newnuc_h2so4_conc_optaa &
+                                        , nait, nacc, max_agepair, n_agepair
+   use modal_aero_amicphys_diags,   only: nqtendaa, nqqcwtendaa &
+                                        , iqtend_cond, iqtend_rnam, iqtend_nnuc, iqtend_coag, iqqcwtend_rnam
+
+   use modal_aero_data,   only: n_mode=>ntot_amode
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/1subarea.F90:56]: is `max_mode` different from `ntot_amode` for MAM4?

File: components/eam/src/chemistry/modal_aero/1subarea.F90 [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+           n_mode,                                        &
+           qgas_cur,          qgas_avg,                   &
+           qgas_netprod_otrproc,                          &
+           qaer_cur,                                      &
+           qnum_cur,                                      &
+           qwtr_cur,                                      &
+           dgn_a,             dgn_awet,         wetdens,  &
+           uptkaer,           uptkrate_h2so4              )
+
+         if (newnuc_h2so4_conc_optaa == 11) then
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/1subarea.F90]: `newnuc_h2so4_conc_optaa` is set to 2 in amicphys_control. I don't think it changes during the simulation. We can make it a parameter and remove codes where it is not 2.

**Review by singhbalwinder — CHANGES_REQUESTED** [review]
Very nice refactor! The comments you added are very useful in following the code and code itself is refactored in a way that it is easy to read. Thanks for cleaning up the code!
I have left some very minor comments.

File: components/eam/src/chemistry/modal_aero/1subarea.F90 (hunk 2 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/1subarea.F90:28]: Thanks!

File: components/eam/src/chemistry/modal_aero/1subarea.F90 (hunk 1 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/1subarea.F90:6]: The different TMRs have different units. I've now noted them later in the code (when the variables are declared). See [1f569df](https://github.com/eagles-project/e3sm_mam4_refactor/pull/56/commits/1f569df66de852defc89e8557d348070b0529eb6).

File: components/eam/src/chemistry/modal_aero/1subarea.F90 (hunk 3 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/1subarea.F90]: Ah, sorry for overlooking this. Updated in [68924d3](https://github.com/eagles-project/e3sm_mam4_refactor/pull/56/commits/68924d3b748cd9184276ae82558254c7ddff56a1).

File: components/eam/src/chemistry/modal_aero/1subarea.F90 (hunk 4 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/1subarea.F90:56]: I printed their values and got 
```
max_mode = 5
ntot_amode = 4
```

File: components/eam/src/chemistry/modal_aero/1subarea.F90 (hunk 5 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/1subarea.F90]: Good point. `newnuc_h2so4_conc_optaa` values other than 2 were probably used only in CAM and during the development of EAMv1. I've updated the code as suggested. See [925c581](https://github.com/eagles-project/e3sm_mam4_refactor/pull/56/commits/925c5814b7a401dc4ee965c783f626e091f24bda).

**Comment by huiwanpnnl** [comment]
_Testing on Compy_

Commands used to run all tests:
```
cd /qfs/people/wanh895/codes/e3sm_mam4_refactor_amicphys_1subarea/
bash mam4_refactor_scripts/compare_baselines_all_tests.sh  \
     -b huiwanpnnl/refactor-amicphys-1subarea_rebased-20230106 \
     -t ALL_TESTS_refactor-amicphys-1subarea_rebased-20230106
```

Commands used to check the results:
```
cd /compyfs/wanh895/e3sm_scratch//ALL_TESTS_refactor-amicphys-1subarea_rebased-20230106
./cs.status.comp_ALL_TESTS_refactor-amicphys-1subarea_rebased-20230106_mam4_org_v2_baselines -s
```

The results were
```
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  DIFF SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel
```

**Comment by huiwanpnnl** [comment]
If I understood it correctly, results from a year-long simulations do not match the baseline. 

The cprnc output in `/compyfs/wanh895/e3sm_scratch/ALL_TESTS_refactor-amicphys-1subarea_rebased-20230106/SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel.C.comp_ALL_TESTS_refactor-amicphys-1subarea_rebased-20230106_mam4_org_v2_baselines/SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel.C.comp_ALL_TESTS_refactor-amicphys-1subarea_rebased-20230106_mam4_org_v2_baselines.eam.h0.0001-12.nc.cprnc.out` reveals that there are 10 out of 416 variables showing differences. Those variables have the suffix `_sfgaex1`, but some other variables with the same suffix show no differences. 

I suspect I had a small bug related to the Fortran arrays used for history output. I'll investigate.

**Comment by singhbalwinder** [comment]
Thanks, Hui for addressing the comments and testing. Typically, if the 1-year simulation is failing, it might fail for a 1-month or shorter run as well. I have put some notes [on confluence](https://eagles-project.atlassian.net/wiki/spaces/Computation/pages/1664253957/Tracking+down+non-BFB+Errors) which might be helpful in setting up a shorter simulation and tracking it down. Let me know if you have any questions on this.

**Comment by huiwanpnnl** [comment]
Yes, @singhbalwinder, I saw your very useful instructions - and will follow them to do my investigation. Thanks!

**Comment by huiwanpnnl** [comment]
Hi @singhbalwinder, sorry for the delay in addressing the nonBFB issue and thanks again for your very helpful instructions on Confluence. My problem was indeed reproducible with 1-month runs and only in the diagnostics.

I resolved the nonBFB issue in [685afa2](https://github.com/eagles-project/e3sm_mam4_refactor/pull/56/commits/685afa254964baeb18cac50c9853923b19090d53). Here are the test results:

*Testing on Compy*

Commands used to run all tests:

```
bash mam4_refactor_scripts/compare_baselines_all_tests.sh \
     -b huiwanpnnl/refactor-amicphys-1subarea_rebased-20230106 \
     -t ALL_TESTS_refactor-amicphys-1subarea_revised-20230123
```

Commands used to check the results:

```
cd /compyfs/wanh895/e3sm_scratch//ALL_TESTS_refactor-amicphys-1subarea_revised-20230123
./cs.status.comp_ALL_TESTS_refactor-amicphys-1subarea_revised-20230123_mam4_org_v2_baselines -s
```

The results were
```
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel
```

**Comment by singhbalwinder** [comment]
Thanks, Hui! I know sometimes it takes a lot of time to track these issues. Everything looks great! I will merge this PR.

**Review by singhbalwinder — APPROVED** [review]

**Pull request merged**